ycmd/completers/typescript/tsserver_completer.py

```diff
@@ -0,0 +1,197 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with YouCompleteMe.  If not, see <http://www.gnu.org/licenses/>.
+
+import json
+import logging
+import os
+import subprocess
+
+from tempfile import NamedTemporaryFile
+
+from ycmd import responses
+from ycmd import utils
+from ycmd.completers.completer import Completer
+
+_logger = logging.getLogger( __name__ )
+
+
+class TSServerCompleter( Completer ):
+  """
+  Completer for TypeScript.
+
+  It uses TSServer which is bundled with TypeScript 1.5
+
+  See the protocol here:
+  https://github.com/Microsoft/TypeScript/blob/2cb0dfd99dc2896958b75e44303d8a7a32e5dc33/src/server/protocol.d.ts
+  """
+
+  def __init__( self, user_options ):
+    super( TSServerCompleter, self ).__init__( user_options )
+
+    binarypath = utils.PathToFirstExistingExecutable( [ 'tsserver' ] )
+    if not binarypath:
+        raise RuntimeError( 'tsserver not found. TypeScript 1.5 or higher is required' )
+
+    self.seq = 0
+    self.proc = utils.SafePopen( binarypath,
+      stdout = subprocess.PIPE,
+      stdin  = subprocess.PIPE
+    )
+
+    _logger.info( 'Enabling typescript completion' )
+
+  def _SendRequest( self, command, arguments=None ):
+    """ Send a request message to TSServer """
+
+    # The server responses contain the sequence
+    # that was sent with the request that triggered it
+    seq = self.seq
+    self.seq += 1
+    request = {
+      'seq':     seq,
+      'type':    'request',
+      'command': command
+    }
+    if arguments:
+      request[ 'arguments' ] = arguments
+    self.proc.stdin.write( json.dumps( request ) )
+    self.proc.stdin.write( "\n" )
+    return seq
+
+  def _ReadResponse( self, seq ):
+    """ Read a response message from TSServer """
+
+    # The headers are pretty similar to HTTP
+    # 'Content-Length' is the only provided header at the time of writing
+    headers = {}
+    while True:
+      headerline = self.proc.stdout.readline().strip()
+      if headerline == "":
+        break
+      [ key, value ] = headerline.split( ':', 1 )
+      headers[ key.strip() ] = value.strip()
+
+    # The response message is a json object which always seems to be on one line.
+    # Since this might change in the future, use the 'Content-Length' header.
+    if 'Content-Length' not in headers:
+      raise RuntimeError( "Missing 'Content-Length' header" )
+    contentlength = int( headers[ 'Content-Length' ] )
+    message = json.loads( self.proc.stdout.read( contentlength ) )
+
+    # The tsserver can sends events, I don't know if they're useful
+    msgtype = message[ 'type' ]
+    if msgtype == 'event':
+      self._HandleEvent( message )
+      return self._ReadResponse()
+
+    if msgtype != 'response':
+      raise RuntimeError( 'Unsuported message type {}'.format( msgtype ) )
```

Comment:
If I'm not mistaken the minimal python supported is 2.6 so positional arguments are required:

```python
raise RuntimeError( 'Unsuported message type {0}'.format( msgtype ) )
```